Readme.md

# Action Toggle Button

simple 2 state button toggling after succesful ajax response

## Installation

Install with [component(1)](http://component.io):

```sh
$ component install chameleonui/action-toggle-button
```

If you want to `build example`, you will need to have stylus and chameleon-stylus-plugin installed.

```sh
$ npm install stylus -g
$ npm install
``` 

## API

```js
new ActionToggleButton([options]);
```

### Options

```jss
method: 'POST',
callback: function (xhrResponse) {
    return xhrResponse.status >= 200 && xhrResponse.status <= 399;
},
states: {
    stateA: {
        content: '<i class="fa fa-star-o"></i>',
        classes: 'button-default',
        nextState: 'stateB',
        title: 'State A'
    },
    stateB: {
        content: '<i class="fa fa-star"></i>',
        classes: 'button-primary',
        nextState: 'stateA',
        title: 'State B'
    },
    XHRrequest: {
        content: '<i class="fa fa-spinner fa-spin"></i>',
        classes: 'button-disabled',
        title: 'Loading ...'
    }
}
```

* **callback** - Function returning boolean true|false. xhrResponse is full response, xhrResponse.status is returned status 


## ToDo

Write complete readme


## Author(s)

[Edgedesign s.r.o.](http://www.edgedesing.cz) – [Daniel Sitek](https://github.com/danielsitek)

## License

The MIT License (MIT)

Copyright © 2013 Edgedesign s.r.o.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

History.md
## v1

### 1.0.2

* Fixed jquery component dependency.
* Fixed link to apiary.io in example.html

### 1.0.1

* Fixed bug for false response status. Now after sending request and getting false from response statuch check, button will reset to it's previous state

### 1.0.0

* First public release

index.js
/**
 * 
 * Action - Toggle Button
 * 
 */

var $ = require('jquery');
var Emitter = require('emitter');

var defaults = {
    method: 'POST',
    callback: function (xhrResponse) {
        return xhrResponse.status >= 200 && xhrResponse.status <= 399;
    },
    states: {
        stateA: {
            content: '<i class="fa fa-star-o"></i>',
            classes: 'button-default',
            nextState: 'stateB',
            title: 'State A'
        },
        stateB: {
            content: '<i class="fa fa-star"></i>',
            classes: 'button-primary',
            nextState: 'stateA',
            title: 'State B'
        },
        XHRrequest: {
            content: '<i class="fa fa-spinner fa-spin"></i>',
            classes: 'button-disabled',
            title: 'Loading ...'
        }
    }
};

function ActionToggleButton(options) {
    this.options = options || {};
    for (var i in defaults) {
        if (!(this.options[i])) this.options[i] = defaults[i];
    }
    this.response = {};
    this.self;
    this.active = 0;
    this.currentState;
    this.nextState;
    this.prevState;
}

module.exports = ActionToggleButton;

Emitter(ActionToggleButton.prototype);

ActionToggleButton.prototype.init = function(self) {
    
    var component = this;
    component.self = self;

    if (!component.currentState) {
        component.currentState = component.self.getAttribute('data-state') ? component.self.getAttribute('data-state') : 'stateA';
    }

    if (!component.nextState) {
        component.nextState = component.options.states[component.currentState] ? component.options.states[component.currentState].nextState : 'stateB';
    }

    if (!component.prevState) {
        component.prevState = component.currentState;
    }

    component._stateChange('XHRrequest');
};

ActionToggleButton.prototype._click = function(fn) {
    var component = this;

    $(component.self).on('click.actiontoggle', function(event) {

        event.preventDefault();
        fn(event);
    });
};

ActionToggleButton.prototype._clearEvent = function() {
    
    var component = this;

    $(component.self).off('click.actiontoggle');
};

ActionToggleButton.prototype._stateChange = function(newStateName) {
    
    var component = this,
        previousStateClasses,
        newStateClasses;

    // clear binded event - prevents multiple clicks
    component._clearEvent();

    // replace full class list
    if (component.options.states[newStateName]) {

        previousStateClasses = component.options.states[component.currentState].classes;
        newStateClasses = component.options.states[newStateName].classes;

        component.self.setAttribute('class', newStateClasses);
        component.self.setAttribute('title', component.options.states[newStateName].title);
        component.self.innerHTML = component.options.states[newStateName].content;

        component.currentState = newStateName;
        component[newStateName]();

    } else {

        console.error('State: ' + newStateName + ' does not exist');
    }
};

ActionToggleButton.prototype.stateA = function() {

    var component = this;
    component.nextState = 'stateB';
    component.prevState = 'stateA';

    component._click(function(){

        component._stateChange('XHRrequest');

    });
};

ActionToggleButton.prototype.stateB = function() {

    var component = this;
    component.nextState = 'stateA';
    component.prevState = 'stateB';

    component._click(function(){

        component._stateChange('XHRrequest');

    });
};

ActionToggleButton.prototype.XHRrequest = function() {
    
    var component = this;

    component._click(function(){});

    component.active = 1;

    component.emit('request');

    component._xhr(component.self.getAttribute('href'), function(xhr){

        component.response = xhr;
        component._XHRresponse();

    });
};

ActionToggleButton.prototype._XHRresponse = function() {
    
    var component = this;

    component.active = 0;
    
    if ( component.options.callback(component.response) ) {

        component._stateChange(component.nextState ? component.nextState : 'stateB');
        component.emit('success');

    } else {

        component._stateChange(component.prevState);
        component.emit('error');
    }

    component.emit('response');
};

ActionToggleButton.prototype._xhr = function(setUrl, resFn) {

    $.ajax({
        url: setUrl
    })
    .done(function(data, textStatus, jqXHR) {
        resFn(jqXHR);
    })
    .fail(function(jqXHR, textStatus, errorThrown) {
        resFn(jqXHR);
    });
};
